### PR TITLE
Support env vars in fixtures

### PR DESCRIPTION
**Pull request by thorsten-stripe**
### Reviewers
r? @tomer-stripe 
cc @stripe/dev-platform

 ### Summary
There is a use case where I might want to set up a customer object with a specific email address or phone number and I want to pull that data from the `.env` file:

```
{
  "_meta": {
    "template_version": 0
  },
  "fixtures": [
    {
      "name": "cust_bender",
      "path": "/v1/customers",
      "method": "post",
      "params": {
        "source": "tok_amex",
        "name": "Bender Bending Rodriguez",
        "email": "bender@planex.com",
        "phone": "${.env:PHONE}",
        "address": {
          "line1": "1 Planet Express St",
          "city": "New New York"
        }
      }
    }
  ]
}
```

I'm not a go expert, and specifically am not sure what `os.Getwd()` 
https://github.com/stripe/stripe-cli/compare/thor/support-env-vars-in-fixtures?expand=1#diff-133d387253cca3c9f0d5be8b5c845987R280
will refer to once compiled. Will that be the directory I'm running `stripe` in?

### CODE REVIEW

File: pkg/fixtures/fixtures.go [hunk 1]
```diff
@@ -275,6 +275,24 @@ func (fxt *Fixture) parseQuery(value string) string {
 		nameAndQuery := r.FindStringSubmatch(value)
 		name := nameAndQuery[1]
 
+		// Catch and insert .env values
+		if name == ".env" {
+			dir, err := os.Getwd() // which dir is this once compiled?
```
Comment (tomer-stripe) [inline pkg/fixtures/fixtures.go]: It'll be the dir you execute from.

**Comment by thorsten-stripe** [comment]
> Generally this looks fine. My one big question is what happens if you don't have an env var. It seems like we need some way to gracefully handle defaults or empty values.

Thanks @tomer-stripe I believe it does error gracefully, as in it outputs the error message but continues with the requests.
image
 
In the case of error it simply returns the string value:
image

**Comment by thorsten-stripe** [comment]
@tomer-stripe any idea what's going wrong in `make ci`?

**Comment by tomer-stripe** [comment]
Hmmm, I'm wondering if every trigger should have some default value so that the user doesn't have to think about it. I think what would be great is if the fixture could act as the default and we could somehow override individual fields in there (check out #295). @ob-stripe suggests using flags but I wonder if env-vars would be easier since we wouldn't need to dynamically support arbitrary flags.

As for the CI, it looks like you added `stripe-go` as a dependency in this PR which we don't use. Try removing that?

**Comment by thorsten-stripe** [comment]
We could support the nullish coalescing operator:
```
"phone": "${.env:PHONE} ?? +1234567890",
```

Or do you think that gets too verbose?

**Comment by tomer-stripe** [comment]
Let me think about this a bit more...

@ob-stripe do you have any ideas here?

**Comment by ob-stripe** [comment]
Hm, I'd prefer not to make the string interpolation too complex, but we could support default values in case the left-hand side is not set (regardless of whether it's a "regular" variable or an environment variable).

Maybe something like this?

```
"phone": "${.env:PHONE|+1234567890}",
```

**Comment by thorsten-stripe** [comment]
@ob-stripe works for me :) I've updated the regex to support default value after pipe. I've also changed from `Find` to `FindR`(https://github.com/thedevsaddam/gojsonq/wiki/Queries#findrpath) to gracefully error when the json_path is incorrect. This means we can now do:

```
"phone": "${.env:PHONE|+1234567890}",
"customer": "${test_customer:id|cus_1234}"
```

File: pkg/fixtures/fixtures.go [hunk 2]
```diff
@@ -269,18 +269,52 @@ func (fxt *Fixture) parseArray(params []interface{}, parent string, index int) [
...
-	r := regexp.MustCompile(`\${(.+):(.+)}`)
+	// to make sure that we're trying to parse a query.
+	// Additionally look for an optional default value: ${name:json_path|default_value}
+	r := regexp.MustCompile(`\${([^\|]+):([^\|]+)\|?(.+)?}`)
 	if r.Match([]byte(value)) {
 		nameAndQuery := r.FindStringSubmatch(value)
 		name := nameAndQuery[1]
 
+		// Check if there is a default value specified
+		if nameAndQuery[3] != "" {
```
Comment (tomer-stripe) [inline pkg/fixtures/fixtures.go]: Do you need to check the length here? (want to make sure we don't get any array out of bounds errors)

File: pkg/fixtures/fixtures.go [hunk 3]
```diff
@@ -269,18 +269,52 @@ func (fxt *Fixture) parseArray(params []interface{}, parent string, index int) [
...
+			value = nameAndQuery[3]
+		}
+
+		// Catch and insert .env values
+		if name == ".env" {
```
Comment (tomer-stripe) [inline pkg/fixtures/fixtures.go]: Are there any unit tests you could add for this? It'd be helpful to verify that the parsing works and we don't accidentally introduce regressions. Check out https://github.com/stripe/stripe-cli/blob/master/pkg/fixtures/fixtures_test.go

File: pkg/fixtures/fixtures.go (hunk 2 above)
Comment (thorsten-stripe) [reply pkg/fixtures/fixtures.go]: @tomer-stripe AFAIK the regex always returns three groups (https://play.golang.org/p/pf_oViIZBah), so we should be good unless we're concerned that this behaviour might change in Go?

**Comment by thorsten-stripe** [comment]
@tomer-stripe I've added test for all cases and they run fine except for `TestParseWithEnvFile` where I'm struggling to write the `.env` file as part of the test. Any idea how to achieve that?

**Comment by tomer-stripe** [comment]
Looks good, ty!

**Pull request merged**